core/iwasm/libraries/libc-builtin/libc_builtin_wrapper.c

```diff
@@ -77,102 +70,19 @@ typedef char *_va_list;
 
-/**
- * @brief Output an unsigned int in hex format
- *
- * Output an unsigned int on output installed by platform at init time. Should
- * be able to handle an unsigned int of any size, 32 or 64 bit.
- * @param num Number to output
- *
- * @return N/A
- */
-static void
-_printf_hex_uint(out_func_t out, void *ctx, const uint64 num, bool is_u64,
-                 enum pad_type padding, int min_width)
-{
-    int shift = sizeof(num) * 8;
-    int found_largest_digit = 0;
-    int remaining = 16; /* 16 digits max */
-    int digits = 0;
-    char nibble;
-
-    while (shift >= 4) {
-        shift -= 4;
-        nibble = (num >> shift) & 0xf;
-
-        if (nibble || found_largest_digit || shift == 0) {
-            found_largest_digit = 1;
-            nibble = (char)(nibble + (nibble > 9 ? 87 : 48));
-            out((int)nibble, ctx);
-            digits++;
-            continue;
-        }
-
-        if (remaining-- <= min_width) {
-            if (padding == PAD_ZERO_BEFORE) {
-                out('0', ctx);
-            }
-            else if (padding == PAD_SPACE_BEFORE) {
-                out(' ', ctx);
-            }
-        }
-    }
-
-    if (padding == PAD_SPACE_AFTER) {
-        remaining = min_width * 2 - digits;
-        while (remaining-- > 0) {
-            out(' ', ctx);
-        }
-    }
-}
-
-/**
- * @brief Output an unsigned int in decimal format
- *
- * Output an unsigned int on output installed by platform at init time. Only
- * works with 32-bit values.
- * @param num Number to output
- *
- * @return N/A
- */
-static void
-_printf_dec_uint(out_func_t out, void *ctx, const uint32 num,
-                 enum pad_type padding, int min_width)
-{
-    uint32 pos = 999999999;
-    uint32 remainder = num;
-    int found_largest_digit = 0;
-    int remaining = 10; /* 10 digits max */
-    int digits = 1;
-
-    /* make sure we don't skip if value is zero */
-    if (min_width <= 0) {
-        min_width = 1;
-    }
-
-    while (pos >= 9) {
-        if (found_largest_digit || remainder > pos) {
-            found_largest_digit = 1;
-            out((int)((remainder / (pos + 1)) + 48), ctx);
-            digits++;
-        }
-        else if (remaining <= min_width && padding < PAD_SPACE_AFTER) {
-            out((int)(padding == PAD_ZERO_BEFORE ? '0' : ' '), ctx);
-            digits++;
-        }
-        remaining--;
-        remainder %= (pos + 1);
-        pos /= 10;
-    }
-    out((int)(remainder + 48), ctx);
-
-    if (padding == PAD_SPACE_AFTER) {
-        remaining = min_width - digits;
-        while (remaining-- > 0) {
-            out(' ', ctx);
-        }
-    }
-}
+#define PREPARE_TEMP_FORMAT()                               \
+    char buf[64], temp_fmt[32], *s;                         \
+                                                            \
+    memset(temp_fmt, 0, sizeof(temp_fmt));                  \
+    bh_memcpy_s(temp_fmt, sizeof(temp_fmt), fmt_start_addr, \
+                fmt - fmt_start_addr + 1);
```

Comment:
fmt may be out of range? Suppose the format specifier is long enough, e.g. "%0123456789012345678901234567890123456789d"